File: src/services/GroupMeetingService.ts
import { GroupMeeting } from '../models/GroupMeeting';
import { GroupMeetingRepository } from '../repositories/GroupMeetingRepository';

export class GroupMeetingService {
  constructor(private groupMeetingRepo: GroupMeetingRepository) {}

  async findSuggestions(
    currentUserId: number,
    title: string,
    startTime: Date,
    endTime: Date,
  ): Promise<GroupMeeting[]> {
    return this.groupMeetingRepo.findAllMatching(
      currentUserId,
      title,
      startTime,
      endTime,
    );
  }

  async getMeetingDetails(meetingId: number): Promise<GroupMeeting | null> {
    return this.groupMeetingRepo.getById(meetingId);
  }

  async joinMeeting(userId: number, meetingId: number): Promise<void> {
    await this.groupMeetingRepo.addParticipant(userId, meetingId);
  }

  async requestJoin(userId: number, meetingId: number): Promise<void> {
    await this.groupMeetingRepo.addRequestToJoin(userId, meetingId);
  }

  async approveUser(ownerId: number, meetingId: number, userId: number): Promise<void> {
    const meeting = await this.groupMeetingRepo.getById(meetingId);

    if (meeting && meeting.ownerId === ownerId) {
      await this.groupMeetingRepo.approveParticipant(userId, meetingId);
    }
  }

  async rejectUser(ownerId: number, meetingId: number, userId: number): Promise<void> {
    const meeting = await this.groupMeetingRepo.getById(meetingId);

    if (meeting && meeting.ownerId === ownerId) {
      await this.groupMeetingRepo.rejectParticipant(userId, meetingId);
    }
  }
}

File: CalendarApi/Services/GroupMeetingService.cs
using CalendarApi.Models;
using CalendarApi.Repositories;

namespace CalendarApi.Services;

/// <summary>
/// Business logic for GroupMeeting operations.
/// Migrated from TypeScript GroupMeetingService.ts
/// </summary>
public class GroupMeetingService
{
    private readonly GroupMeetingRepository _groupMeetingRepo;

    public GroupMeetingService(GroupMeetingRepository groupMeetingRepo)
    {
        _groupMeetingRepo = groupMeetingRepo;
    }

    public List<GroupMeeting> FindSuggestions(int currentUserId, string title, DateTime startTime, DateTime endTime)
    {
        return _groupMeetingRepo.FindAllMatching(currentUserId, title, startTime, endTime);
    }

    public GroupMeeting? GetMeetingDetails(int meetingId)
    {
        return _groupMeetingRepo.GetById(meetingId);
    }

    public void JoinMeeting(int userId, int meetingId)
    {
        _groupMeetingRepo.AddParticipant(userId, meetingId);
    }

    public void RequestJoin(int userId, int meetingId)
    {
        _groupMeetingRepo.AddRequestToJoin(userId, meetingId);
    }

    public void ApproveUser(int ownerId, int meetingId, int userId)
    {
        var meeting = _groupMeetingRepo.GetById(meetingId);
        if (meeting != null && meeting.OwnerId == ownerId)
        {
            _groupMeetingRepo.ApproveParticipant(userId, meetingId);
        }
    }

    public void RejectUser(int ownerId, int meetingId, int userId)
    {
        var meeting = _groupMeetingRepo.GetById(meetingId);
        if (meeting != null && meeting.OwnerId == ownerId)
        {
            _groupMeetingRepo.RejectParticipant(userId, meetingId);
        }
    }
}
